### ISSUE TITLE
FacetedSearch on nested document

### ISSUE THREAD

**Issue by yml** [elastic/elasticsearch-dsl-py#625]
The goal is to be able to do faceted search on a date (departure_date) store on a nested document. 
I am using elasticsearch 5.1.2, elasticsearch-py==5.2.0, elasticsearch-dsl-py==5.1.0.

The simplified DocType look like this:

```
class TourDoc(DocType):
    title = Text(analyzer="english")
    countries = Keyword()
    regions = Keyword()
    departures = Nested(
        doc_class=DepartureDoc,
        properties={
            "departure_date": Date(),
            "returning_date": Date(),
        }
    )
```

The FacedtedSearch sub class : 

```
class TourSearch(FacetedSearch):
    doc_types = [TourDoc,]
    fields = ("title",)

    facets = {
        'regions': TermsFacet(field='regions'),
        'countries': TermsFacet(field='countries'),
        'departure_date': DateHistogramFacet(
            field='departures.departure_date',
            interval='month',
            min_doc_count=0,
        )
    }
```

Here is is the interactive session that illustrates the issue:

```
In [2]: ts = searches.TourSearch()

In [3]: ts._s.to_dict()
Out[3]: 
{'aggs': {'_filter_countries': {'aggs': {'countries': {'terms': {'field': 'countries'}}},
   'filter': {'match_all': {}}},
  '_filter_departure_date': {'aggs': {'departure_date': {'date_histogram': {'field': 'departures.departure_date',
      'interval': 'month',
      'min_doc_count': 0}}},
   'filter': {'match_all': {}}},
  '_filter_regions': {'aggs': {'regions': {'terms': {'field': 'regions'}}},
   'filter': {'match_all': {}}}},
 'highlight': {'fields': {'title': {}}},
 'query': {'match_all': {}}}

In [4]: results= ts.execute()

In [5]: results.facets
Out[5]: {'regions': [(u'Europe', 119, False), (u'U.S. & Canada', 46,...}

In [6]: results.facets.__dict__
Out[6]: 
{'_d_': {'countries': [(u'United States', 41, False),
   (u'Italy', 30, False),
   (u'France', 27, False),
   (u'United Kingdom', 16, False),
   (u'England', 15, False),
   (u'Spain', 14, False),
   (u'Mediterranean', 13, False),
   (u'Greece', 12, False),
   (u'Germany', 10, False),
   (u'Cambodia', 7, False)],
  'departure_date': [],
  'regions': [(u'Europe', 119, False),
   (u'U.S. & Canada', 46, False),
   (u'Asia', 30, False),
   (u'Africa', 21, False),
   (u'Central & South America', 20, False),
   (u'Australia & the Pacific', 12, False),
   (u'Polar', 1, False)]}}
```

I have also tried to replace the dotted definition of the nested document by a double underscored version  `field='departures__departure_date',` without success

In  search for a solution I tried to patch the generated dict to make it aware of the nested nature of my dates:

```
In [2]: ts = searches.TourSearch()

In [3]: query_dict = ts._s.to_dict()

In [4]: query_dict['aggs']['_filter_departure_date']['nested']  = {"path": "departures"}

In [5]: query_dict
Out[5]: 
{'aggs': {'_filter_countries': {'aggs': {'countries': {'terms': {'field': 'countries'}}},
   'filter': {'match_all': {}}},
  '_filter_departure_date': {'aggs': {'departure_date': {'date_histogram': {'field': 'departures.departure_date',
      'interval': 'month',
      'min_doc_count': 0}}},
   'filter': {'match_all': {}},
   'nested': {'path': 'departures'}},
  '_filter_regions': {'aggs': {'regions': {'terms': {'field': 'regions'}}},
   'filter': {'match_all': {}}}},
 'highlight': {'fields': {'title': {}}},
 'query': {'match_all': {}}}
In [9]: del(query_dict['aggs']['_filter_departure_date']['filter']) 

In [21]: ts._s.execute().aggregations.__dict__
Out[21]: 
{'_aggs': AggsProxy(aggs={'_filter_regions': Filter(aggs={'regions': Terms(field='regions')}, filter=MatchAll()), '_filter_departure_date': Filter(aggs={'departure_date': DateHistogram(field='d
epartures.departure_date', interval='month', min_doc_count=0)}, filter=MatchAll()), '_filter_countries': Filter(aggs={'countries': Terms(field='countries')}, filter=MatchAll())}),
 '_d_': {u'_filter_countries': {u'countries': {u'buckets': [{u'doc_count': 41,
      u'key': u'United States'},
     {u'doc_count': 30, u'key': u'Italy'},
     {u'doc_count': 27, u'key': u'France'},
     {u'doc_count': 16, u'key': u'United Kingdom'},
     {u'doc_count': 15, u'key': u'England'},
     {u'doc_count': 14, u'key': u'Spain'},
     {u'doc_count': 13, u'key': u'Mediterranean'},
     {u'doc_count': 12, u'key': u'Greece'},
     {u'doc_count': 10, u'key': u'Germany'},
     {u'doc_count': 7, u'key': u'Cambodia'}],
    u'doc_count_error_upper_bound': 5,
    u'sum_other_doc_count': 269},
   u'doc_count': 225},
  u'_filter_departure_date': {u'departure_date': {u'buckets': []},
   u'doc_count': 225},
  u'_filter_regions': {u'doc_count': 225,
   u'regions': {u'buckets': [{u'doc_count': 119, u'key': u'Europe'},
     {u'doc_count': 46, u'key': u'U.S. & Canada'},
     {u'doc_count': 30, u'key': u'Asia'},
     {u'doc_count': 21, u'key': u'Africa'},
     {u'doc_count': 20, u'key': u'Central & South America'},
     {u'doc_count': 12, u'key': u'Australia & the Pacific'},
     {u'doc_count': 1, u'key': u'Polar'}],
    u'doc_count_error_upper_bound': 0,
    u'sum_other_doc_count': 0}}},
 '_search': <elasticsearch_dsl.search.Search at 0x7f3c886fc910>}

In [23]: searches.TourDoc.search().from_dict(query_dict).execute().aggregations.__dict__
Out[23]: 
{'_aggs': AggsProxy(aggs={'_filter_regions': Filter(aggs={'regions': Terms(field='regions')}, match_all={}), '_filter_departure_date': Nested(aggs={'departure_date': DateHistogram(field='depart
ures.departure_date', interval='month', min_doc_count=0)}, path='departures'), '_filter_countries': Filter(aggs={'countries': Terms(field='countries')}, match_all={})}),
 '_d_': {u'_filter_countries': {u'countries': {u'buckets': [{u'doc_count': 41,
      u'key': u'United States'},
     {u'doc_count': 30, u'key': u'Italy'},
     {u'doc_count': 27, u'key': u'France'},
     {u'doc_count': 16, u'key': u'United Kingdom'},
     {u'doc_count': 15, u'key': u'England'},
     {u'doc_count': 14, u'key': u'Spain'},
     {u'doc_count': 13, u'key': u'Mediterranean'},
     {u'doc_count': 12, u'key': u'Greece'},
     {u'doc_count': 10, u'key': u'Germany'},
     {u'doc_count': 7, u'key': u'Cambodia'}],
    u'doc_count_error_upper_bound': 5,
    u'sum_other_doc_count': 269},
   u'doc_count': 225},
  u'_filter_departure_date': {u'departure_date': {u'buckets': [{u'doc_count': 3,
      u'key': 1393632000000,
      u'key_as_string': u'2014-03-01T00:00:00.000Z'},
     {u'doc_count': 2,
      u'key': 1396310400000,
      u'key_as_string': u'2014-04-01T00:00:00.000Z'},
     {u'doc_count': 6,
      u'key': 1398902400000,
      u'key_as_string': u'2014-05-01T00:00:00.000Z'},
     {u'doc_count': 16,
      u'key': 1509494400000,
      u'key_as_string': u'2017-11-01T00:00:00.000Z'},
     {u'doc_count': 6,
      u'key': 1512086400000,
      u'key_as_string': u'2017-12-01T00:00:00.000Z'},
     {u'doc_count': 14,
      u'key': 1514764800000,
      u'key_as_string': u'2018-01-01T00:00:00.000Z'},
     {u'doc_count': 19,
      u'key': 1517443200000,
      u'key_as_string': u'2018-02-01T00:00:00.000Z'},
     {u'doc_count': 6,
      u'key': 1519862400000,
      u'key_as_string': u'2018-03-01T00:00:00.000Z'},
     {u'doc_count': 2,
      u'key': 1522540800000,
      u'key_as_string': u'2018-04-01T00:00:00.000Z'},
     {u'doc_count': 0,
      u'key': 1525132800000,
      u'key_as_string': u'2018-05-01T00:00:00.000Z'},
     {u'doc_count': 1,
      u'key': 1527811200000,
      u'key_as_string': u'2018-06-01T00:00:00.000Z'},
     {u'doc_count': 0,
      u'key': 1530403200000,
      u'key_as_string': u'2018-07-01T00:00:00.000Z'},
     {u'doc_count': 0,
      u'key': 1533081600000,
      u'key_as_string': u'2018-08-01T00:00:00.000Z'},
     {u'doc_count': 1,
      u'key': 1535760000000,
      u'key_as_string': u'2018-09-01T00:00:00.000Z'}]},
   u'doc_count': 539},
  u'_filter_regions': {u'doc_count': 225,
   u'regions': {u'buckets': [{u'doc_count': 119, u'key': u'Europe'},
     {u'doc_count': 46, u'key': u'U.S. & Canada'},
     {u'doc_count': 30, u'key': u'Asia'},
     {u'doc_count': 21, u'key': u'Africa'},
     {u'doc_count': 20, u'key': u'Central & South America'},
     {u'doc_count': 12, u'key': u'Australia & the Pacific'},
     {u'doc_count': 1, u'key': u'Polar'}],
    u'doc_count_error_upper_bound': 0,
    u'sum_other_doc_count': 0}}},
 '_search': <elasticsearch_dsl.search.Search at 0x7f3c88521210>}
``` 
The problem I am facing is when I try to replace `ts._s ` with something that is aware of the nesting. I tried `ts._s.from_dict(query_dict)` but t does not seems to work.

Is this problem known ? Is there a simple fix at the user level or does the faceted search should not be used on nested document.

**Comment by yml** [comment]
```

**Comment by HonzaKral** [comment]
Currently `FacetedSearch` doesn't support nested documents as the code would get too complex for the generic use case. What I can recommend to you is creating a custom `Facet` subclass that will generate the proper aggregation as well as filter.

I am sorry for the complications!

**Comment by OskarPersson** [comment]
Is there a documented example of a custom Facet subclass that works with nested documents?

**Comment by HonzaKral** [comment]
Thank you @OskarPersson for your question, it made me realize I was wrong earlier when I assumed that `NestedFacet` would get too complex so I implemented it in 97e3f75

Sorry, @yml, for the earlier error!
